File: crates/aether-evals/src/agents/docker_agent.rs
use super::transcript::is_terminal;
use super::{Agent, AgentRunResult, RunError};
use crate::Task;
use crate::containers::{Container, ContainerError};
use aether_core::events::AgentEvent;
use async_stream::try_stream;
use futures::Stream;
use std::collections::{BTreeMap, HashMap};
use tokio::io::AsyncBufReadExt;

/// Environment variable through which `DockerAgent` hands the caller-provided task prompt to the
/// in-container command.
pub const AETHER_EVAL_TASK_PROMPT_ENV: &str = "AETHER_EVAL_TASK_PROMPT";
pub const AETHER_EVAL_WORKSPACE_ROOT_ENV: &str = "AETHER_EVAL_WORKSPACE_ROOT";
pub const AETHER_EVAL_CWD_ENV: &str = "AETHER_EVAL_CWD";

pub struct DockerAgent {
    container: Container,
    command: Vec<String>,
    env_vars: HashMap<String, String>,
}

impl DockerAgent {
    /// `command` is the in-container argv to exec; its stdout must be newline-delimited
    /// `AgentEvent` JSON.
    pub fn new(container: Container, command: Vec<String>) -> Self {
        Self { container, command, env_vars: HashMap::new() }
    }

    pub fn with_env_var(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.env_vars.insert(key.into(), value.into());
        self
    }

    pub fn with_env_vars(mut self, env_vars: impl IntoIterator<Item = (String, String)>) -> Self {
        self.env_vars.extend(env_vars);
        self
    }
}

impl Agent for DockerAgent {
    fn run(&self, task: Task) -> impl Stream<Item = AgentRunResult> + Send + '_ {
        let env = {
            let mut env: BTreeMap<_, _> = self.env_vars.iter().map(|(k, v)| (k.clone(), v.clone())).collect();
            env.insert(AETHER_EVAL_TASK_PROMPT_ENV.to_string(), task.prompt().to_string());
            env.insert(AETHER_EVAL_CWD_ENV.to_string(), self.container.cwd().display().to_string());
            env.insert(
                AETHER_EVAL_WORKSPACE_ROOT_ENV.to_string(),
                self.container.workspace_root().display().to_string(),
            );
            env
        };

        try_stream! {
            let mut exec = self.container.exec_streaming(self.command.clone(), env).await?;
            let finished = {
                let mut lines = exec.stdout().lines();
                let mut finished = false;
                while let Some(line) = lines.next_line().await.map_err(|source| ContainerError::StdoutRead { source })? {
                    if line.trim().is_empty() {
                        continue;
                    }
                    tracing::debug!("docker agent stdout: {line}");
                    let message: AgentEvent = serde_json::from_str(&line)
                        .map_err(|source| RunError::AgentEventJsonLine { line, source })?;
                    let terminal = is_terminal(&message);
                    yield message;
                    if terminal {
                        finished = true;
                        break;
                    }
                }
                finished
            };

            let stderr = exec.stderr_to_string().await?;
            if !stderr.trim().is_empty() {
                tracing::debug!("docker agent stderr: {}", stderr.trim());
            }

            if !finished {
                Err::<AgentEvent, RunError>(RunError::CommandExitWithoutTerminal { stderr })?;
            }
        }
    }
}

File: packages/aether-evals/src/DockerAgent.ts
import { env as processEnv } from "node:process";
import type { AgentEvent } from "@aether-agent/sdk";
import { Container } from "./containers/container.js";
import { AetherEvalError } from "./errors.js";
import type { Agent } from "./Agent.js";
import type { Task } from "./task.js";
import { isTerminalEvent } from "./transcript.js";

export const CONTAINER_AETHER_HOME = "/root/.aether";
export const AETHER_EVAL_TASK_PROMPT_ENV = "AETHER_EVAL_TASK_PROMPT";
export const AETHER_EVAL_WORKSPACE_ROOT_ENV = "AETHER_EVAL_WORKSPACE_ROOT";
export const AETHER_EVAL_CWD_ENV = "AETHER_EVAL_CWD";
const PROVIDER_ENV_VARS = new Set([
  "ANTHROPIC_API_KEY",
  "DEEPSEEK_API_KEY",
  "GEMINI_API_KEY",
  "MOONSHOT_API_KEY",
  "OPENAI_API_KEY",
  "OPENROUTER_API_KEY",
  "ZAI_API_KEY",
]);

export interface DockerAgentOptions {
  container: Container;
  command: string[];
  env?: Record<string, string>;
}

export class DockerAgent implements Agent {
  readonly container: Container;
  readonly command: string[];
  readonly env: Record<string, string>;

  constructor(options: DockerAgentOptions) {
    if (options.command.length === 0) {
      throw new AetherEvalError(
        "configuration_error",
        "DockerAgent command must not be empty",
      );
    }
    this.container = options.container;
    this.command = [...options.command];
    this.env = {
      ...defaultEvalEnvVars(processEnv),
      ...(options.env ?? {}),
    };
  }

  withEnvVar(key: string, value: string): DockerAgent {
    return new DockerAgent({
      container: this.container,
      command: this.command,
      env: { ...this.env, [key]: value },
    });
  }

  withEnvVars(env: Record<string, string>): DockerAgent {
    return new DockerAgent({
      container: this.container,
      command: this.command,
      env: { ...this.env, ...env },
    });
  }

  async *run(task: Task): AsyncIterable<AgentEvent> {
    const env = {
      ...this.env,
      [AETHER_EVAL_WORKSPACE_ROOT_ENV]: this.container.workspaceRoot,
      [AETHER_EVAL_CWD_ENV]: this.container.cwd,
      [AETHER_EVAL_TASK_PROMPT_ENV]: task.prompt,
    };

    let stderr = "";
    let finished = false;
    for await (const line of this.container.execStreaming({
      command: this.command,
      cwd: this.container.cwd,
      env,
      onStderr: (chunk) => {
        stderr += chunk;
      },
    })) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      const message = parseAgentEvent(trimmed);
      finished = isTerminalEvent(message);
      yield message;
      if (finished) break;
    }

    if (!finished) {
      throw new AetherEvalError(
        "command_exit_without_terminal",
        `agent command exited without emitting a terminal AgentEvent.\nstderr:\n${stderr}`,
      );
    }
  }
}

export function defaultEvalEnvVars(
  hostEnv: Record<string, string | undefined> = processEnv,
): Record<string, string> {
  const env: Record<string, string> = {};
  for (const [key, value] of Object.entries(hostEnv)) {
    if (value === undefined) continue;
    if (key === "AETHER_HOME") continue;
    if (
      PROVIDER_ENV_VARS.has(key) ||
      key === "OLLAMA_HOST" ||
      key.startsWith("AETHER_")
    ) {
      env[key] = value;
    }
  }
  env.AETHER_HOME = CONTAINER_AETHER_HOME;
  return env;
}

function parseAgentEvent(line: string): AgentEvent {
  try {
    return JSON.parse(line) as AgentEvent;
  } catch (err) {
    throw new AetherEvalError(
      "agent_event_json_line",
      `agent emitted an invalid AgentEvent JSON line: ${line}`,
      err,
    );
  }
}
